src/safeposix/syscalls/fs_calls.rs

```diff
@@ -1163,8 +1158,23 @@ impl Cage {
             Pipe(pipe_filedesc_obj) => {
                 pipe_filedesc_obj.pipe.incr_ref(pipe_filedesc_obj.flags);
             }
-            Socket(_socket_filedesc_obj) => {
-                //we handle the closing of these on drop
+            Socket(ref socket_filedesc_obj) => {
+                //we handle the closing of sockets on drop
+                // checking whether this is a domain socket
+
+                let sock_tmp = socket_filedesc_obj.handle.clone();
+                let sockhandle = sock_tmp.write();
+                let socket_type = sockhandle.domain;
+                if socket_type == AF_UNIX {
+                    if let Some(sockinfo) = &sockhandle.unix_info {
+                        if let Some(sendpipe) = sockinfo.sendpipe.as_ref() {
+                            sendpipe.incr_ref(O_WRONLY);
+                        }
+                        if let Some(receivepipe) = sockinfo.receivepipe.as_ref() {
+                            receivepipe.incr_ref(O_RDONLY);
+                        }
```

Comment:
@yizhuoliang

Reply:
Upon further inspection this was correct originally, we want to do operations here on both the send and receive pipes since the socket is modeled as two pipes. Same for the next comment.